File: elixir/word-count/lib/word_count.ex
defmodule WordCount do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence),
    do:
      sentence
      |> String.downcase()
      |> String.replace(~r/[^[:alnum:]-]+/u, " ")
      |> String.split()
      |> Enum.reduce(%{}, fn word, acc ->
        Map.put(acc, word, Map.get(acc, word, 0) + 1)
      end)
end

File: lua/word-count/word-count.lua
local function word_count(s)
    local count = {}
    for word in s:lower():gsub("[\n:!@$%%^&:,.]", " "):gmatch("%S+") do
        word = string.gsub(word, "^'*(.-)'*$", "%1")
        if count[word] == nil then
            count[word] = 1
        else
            count[word] = count[word] + 1
        end
    end
    return count
end

return {
  word_count = word_count,
}
